Architectural Rework

opadron opened the pull request:
This PR will merge in the effort to redesign several major parts of the OSUMO application architecture.  See commit messages for details on the changes.

Commit:
wip - add sample specs

Commit:
wip - spec work

Commit:
add server support for new spec architecture

 - This commit lays the foundation for the new architectural
   changes, and starts to make some of the design ideas more
   concrete, at least on the server side of things.
   Conceptually, there isn't much being added on the server
   side; the primary purpose of this backend work is to
   enable the use of a shorthand notation when submitting
   jobs (see example).

 - No UI work, yet, but I was able to test out the new
   mechanism for submitting jobs using the following code
   snippet:

  ```javascript
  rest({
    path: 'osumo/task_spec/iGPSe/run?title=CustomTitle',
    type: 'POST',
    data: {
      'INPUT(mrna_input_path)':
        'FILE:57a8b6247be3a054f8be9106',
      'INPUT(mirna_input_path)':
        'FILE:57a8b6247be3a054f8be90fe',
      'INPUT(clinical_input_path)':
        'FILE:57a8b6247be3a054f8be9117',
      'INPUT(mrna_clusters)':
        'INTEGER:5',
      'INPUT(mirna_clusters)':
        'INTEGER:5',
      'OUTPUT(output_mrna_dim)':
        'FILE:582bc1517be3a024f80ff17c:custom_filename.csv',
      'OUTPUT(output_mrna_heatmap)':
        'FILE:57bea8f8bb1162473a254e97(format:"jpeg")'
    }
  });
  ```

Commit:
add minor server fixes & tweaks

Commit:
add placeholder process-page component/test scaffolding

Commit:
add analysis reducers and analysis components

Commit:
flesh out field and button analysis element types

 - add new global actions
   - addAnalysisElement
   - addAnalysisPage
   - removeAnalysisElement
   - removeAnalysisPage
   - setAnalysisForm

 - add new components
   - body/analysis/element
   - body/analysis/element/button
   - body/analysis/element/field

 - add temporary test scaffolding that
   automatically runs after a delay at startup

Commit:
start fleshing out file/folder selection elements

 - add property to page spec: "main_action"

   - since every page is rendered as a form, there
     should be some "main" action that takes place
     onSubmit.  This property names the action to
     use.

   - Also, button elements no longer need to
     provide an "action" property and can instead
     default to submitting the form.

   - renamed the "form" page property to "name"

   - add new analysis element component:
     file-selection

   - add initial support for file_selection and
     folder_selection analysis element types

   - add front-end support for more page
     properties
     - description
     - name (formerly "form")
     - notes

   - add new css in top-level analysis component

     - I could use some help cleaning the css
       situation up.  A lot of what's in the file
       was taken from @manthey's CSS, and actually
       had to be further tweaked because it
       clashed with bootstrap in ways I don't
       fully understand.

Commit:
add control flow/analysis components/more cleanup

 - add new npm dependencies
   - file-loader
   - urijs
   - url-loader

 - rename spec keywords to remain consistent with
   the camelCase convention
   - data_type -> dataType
   - file_selection -> fileSelection
   - folder_selection -> folderSelection
   - main_action -> mainAction

 - rework actions.jsx and utils/router.jsx modules
   - I started running into a lot of headaches
     with how the routing was handling the opening
     and closing of the various dialogs.

   - I ended up simplifying the router, and
     switching the part of the route used for
     dialogs so that now the section is a
     psuedo-query that sits at the end of the
     location hash (which is how Girder seems to
     handle it).

   - While I was at it, I also wrapped actions
     that interact with routes so that triggering
     those actions will always trigger a reroute,
     and vice-versa.

   - removed global action "clearDialog"

     - no longer needed since closing a dialog
       will now always trigger a reroute and
       vice-versa.  Just use closeDialog from now
       on.

 - changed global action "setAnalysisForm" into
   new action "setAnalysisFormState".

 - changed analysis reducer "setForm" into
   new reducer "setFormState".

   - These changes are to address the issue with
     handling the states of each analysis element
     while also handling the automatic aggregation
     of the values produced by the elements in a
     page when an action on an analysis page is
     triggered.  For now, each element just has a
     state that's part of the redux state tree,
     and any arbitrary properties can be set on
     that object.  The only special property is
     "value", which is what is consulted when
     building the form on action trigger.

 - Started adding callback and state connection
   logic to the analysis component subtree.

   - The subtree is rooted at the Analysis
     component (see
     components/containers/analysis.jsx).

   - so far, I've identified three top-level
     callbacks:

     - onFileSelect(page, formKey): triggered when
       an analysis element needs to open the file
       selection dialog to make a file selection

     - onStateChange(page, formKey, newState):
       triggered when an analysis element needs to
       update the portion of the form state that
       it is tracking.  This callback is roughly
       equivalent to calling this.setState, in a
       non-redux app.

     - onAction(page, action): triggered when a
       page's action should be dispatched (e.g.: a
       button is clicked or the page's form is
       submitted).  This callback is the one I
       still need to work on the most, but the
       current stub illustrates the idea:
       aggregate the form data for each of the
       page's elements and pass that data, along
       with the page and the action to some global
       registry of actions.  This registry, being
       composed exclusively of unserializable
       functions, would likely be separate from
       the redux state tree.

Commit:
improve action trigger code/rework server logic

 Actions are now working in a limited sense.  The
 code that determines what an "action" does is
 stored in a global action registery.  This
 registery needs to be programmatically created
 and populated up-front -- before actions are
 triggered.  Eventually, it would be nice to make
 that "code" part of the task YAML spec, but
 figuring out how to make that work will probably
 be a lot of work.  For now, we have a
 manually-managed registery.

 Job submission works.  Following up on the
 results of a job is next on the list.

 - moved some variables around in
   AnalysisContainer so that onAction isn't
   recreated on every state change.

 - added new module: utils/analysis

   - added new helper function: aggregateForm

 - added action registery to web app (globals.jsx
   and actions.jsx)

   - added new global actions
     - registerAnalysisAction
     - triggerAnalysisAction

 - modify file-selection analysis element

   - For file selections, the text field is not
     editable.  Use the file-selection dialog, for
     now.

   - For folder selections, the dialog button is
     disabled.  Use the text field to enter the
     folder id directly.

   - These missing features will be added later.

 - Rework server code

   - Added new resource type: jobuser

     - It's basically a (userId, jobId) mapping
       along with the list of processed files.

     - Only jobs scheduled through OSUMO have one
       of these created for them.

     - There might be even better ways to store
       this information, but at least this is an
       improvement over a memory-backed dict in
       that it's now persistent across server
       restarts and doesn't mess with the job
       model directly.

   - Remove some old code that isn't used anymore

     - getTasks()

     - processTask()

   - Current routes:

     - osumo/task -> getTaskSpecs

       - search for a set of task specs

     - osumo/task/:name -> getTaskSpecByName

       - fetch a particular task spec by name

     - osumo/task/:name/run -> runTaskSpec

       - create a job from this task spec

     - osumo/results/:jobId -> getTaskResults

       - return the status of an OSUMO job and, if
         the job had completed successfully, the
         list of files it had created

Commit:
add new page after successful run

For now, I just got the heatmaps to show up, but
I'll need to take a deeper dive into the
parallelsets code to integrate it as an analysis
element type.

 - minor server-side work

   - added task spec: "iGPSePart2"

   - removed some duplicated R code.

 - front-end work

   - added new analysis element type: "image"

   - added new utility functions
     (utils/analysis.jsx):

     - runTask()

     - pollJob()

   - started updating igpse "process" action so
     that it adds a new page when the job finishes
     successfully.

Commit:
WIP - reducer overhaul

Commit:
fix file navigation root selector

Commit:
formally replace old reducers with new ones

Commit:
add new analysis element type: 'parallel-sets'

 - Added new NPM dependencies:
   - d3-array
   - d3-scale
   - d3-shape

 - Added a new analysis element type:
   'parallel-sets'

   - Element is a port + partial rewrite of the
     parallel sets code from Hao that uses the new
     d3 API + React for the DOM management.

 - Factored out code that handled formatting of
   analysis element descriptions, notes, and names
   into a common section in element/index.jsx.

 - Removed most of the temporary staging code in
   main.jsx.

Commit:
migrate plugin to Girder 2.0

Commit:
take advantage of new girder 2.0 features

Updates the client application code to take
advantage of the new features in Girder 2.0.  Now,
file selection and folder selection work as
expected, including automatic client-side
filtering of presented items.  With these change
sets in place the bulk of the architectural rework
is nearly complete.

 - Updated redux state tree shape to hold new
   state information.

   - dialog.fileNavigation is now
     dialog.fileSelect

   - added new fields

     - dialog.fileSelect.folderSelectMode

     - dialog.fileSelect.showItems

 - Addeed new reducers supporting new dispatch
   action types:

   - SET_FILE_NAVIGATION_FOLDER_SELECT_MODE

   - SET_FILE_NAVIGATION_FOLDER_SHOW_ITEMS

 - Added new global action creators

   - setFileNavigationFolderSelectMode

   - setFileNavigationShowItems

   - setItemFilter

 - Simplified file-select and folder-select
   components.

    - Removed "recently chosen" dropdown

    - Unified look and feel

 - Updated analysis item selected callback to
   support items and non-items
   (folders/user/collections)

 - Updated file-selector dialog to support folder
   selection mode, and client-side filtering.

 - Fix several issues with how the HierarchyWidget
   view was being used within the file-selector
   component.

Commit:
add convenience tools for component authors

Includes a miscellaneous set of changes aimed at
producing convenience functions and similar
functionality for making it easier for analysis
component authors to write their components.  Also
includes various fixes.

 - BACKEND changes:

   - added new routes

     - osumo/ui/:key -> getUISpecByKey()

     - osumo/ui -> getUISpecs()

   - fixed an issue where a shorthand STRING value
     would not be parsed correctly.

   - fixed an issue where the Osumo resource was
     not being properly mounted.

   - switched parsing of ui specs so that the key
     of every ui spec is unconditionally set to
     its filename's basename (e.g.: igpse.yml ->
     igpse)

   - added a tags section for ui specs

   - added new ui specs

     - igpse2

     - igpse3

   - (temporarily) removed old job specs/cleaned
     up ui_specs directory

     - kmeans and friends will be reimplemented
       later

 - FRONTEND changes:

   - added new reducer action type:
     "TRUNCATE_ANALYSIS_PAGES"

   - added new global action creator:
     "truncateAnalysisPages"

   - fixed an issue in the analysis reducers where
     a partially constructed parent object is
     operated on.  This issue would cause the
     reducer to return an incorrect new state
     and/or the global action to return an
     incomplete parent.

   - added new section: ./analysis-modules

     - section will contain the coding components
       of the analysis workflows

   - added new analysis modules

     - igpse

     - igpse2

     - igpse3

   - removed debugging text in various analysis
     components

   - added new analysis element type: hidden

     - works as a surprisingly effective "hack"
       for adding elements that may have state,
       but not show anything to the user

     - used in the igpse workflow to shuttle
       transferData along from one step to the
       next.

   - added new analysis utility functions
     (./utils/analysis)

     - processAnalysisPage

     - fetchAndProcessAnalysisPage

Commit:
complete architecture rework and iGPSe workflow

Completes the vast majority of the architectural
rework.  Updates task specs, ui specs, R codes,
and analysis modules so that the origin iGPSe
workflow functions as it had before the rework.

 - BACKEND changes:

   - completely removed job_specs

   - task spec routes modified so that task specs
     are identified by key instead of name.

     - done to become more consistent with ui spec
       handling

     - Osumo.getTaskSpecByName renamed to
       Osumo.getTaskSpecByKey

     - With the added requirement that all ui and
       task specs be identified by a unique key,
       the need for the index parameter is
       eliminated.

   - fixed issues with the iGPSe R code

     - including a work-around for an issue in
       girder-worker that prevents it from
       automatically serializing & deserializing
       data correctly

     - work-around is mirrored in the iGPSe2 R
       code.

     - fixed issue with new version of jsonlite
       that would cause the data returned to
       deviate from what the code had expected.

 - FRONTEND changes:

   - added file:
     ./analysis-modules/base-listing.jsx

   - exports an array of entries: one for each
     "base" module, a "base" module being defined
     as the module that constitutes the "first
     step" in a workflow.

     - array is used to populate the new top-level
       dropdown in the analysis page.

   - fixed an issue in the processAnalysisPage and
     fetchAndProcessAnalysisPage utility functions
     that made it easy to accidentally return
     errorenous data in the preprocess and
     postprocess callbacks.

Commit:
add final tweaks and reimplment other analyses

Adds a few final improvements to the new
architecture.  Adds other analysis workflows from
before the rework.  Brings this branch to
feature-parity with master.

 - BACKEND changes

   - fixed an issue where the "extra_args" section
     of the arg shorthand syntax was not being
     properly parsed for INPUT args.

   - added a new INPUT arg shorthand type:
     "BOOLEAN"

     - takes an integer and coerces to json "true"
       if the integer is non-zero and json
       "false", otherwise

   - reimplemented old workflows to work with the
     new architecture

     - "kmeans"

     - "silhouette"

     - "surv"

 - FRONTEND changes

   - added new analysis modules

     - "kmeans"

     - "silhouete"

     - "silhouete2"

     - "surv"

     - "surv2"

Commit:
fix typo in web_client index page

Commit:
fix travis

Commit:
fix package.json using the old web-external name

Commit:
fix lint errors/remove old code

Commit:
more travis fixing

Commit:
tweaks for demo/tutorial docs

Commit:
fix several minor bugs that caused igpse and other analysis to fail

waxlamp reviewed (approved):
Too much to really review but the application behavior is good and I think we can address any problems in forthcoming PRs as we go.

manthey committed:
Improve travis build.

We weren't building the dev requirements of girder, necessary for running some of the tests.

The R proxy package now only works with a later version of R than travis provides, so install an explicit version.

Preinstall some more of the R packages via ubuntu packages where possible.

Adjust the order of some commands.
